### PR TITLE
HPCC4J-468 Security exploit junit tests

### PR DESCRIPTION
**Pull request by rpastrana**
- Creates XML parsing exploit tests

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [x] I have created a corresponding JIRA ticket for this submission
- [x] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [x] The change has been fully tested:
  - [ ] I have performed unit tests to cover my changes.
  - [ ] I have performed system test and covered possible regressions and side effects.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.

### CODE REVIEW

**Comment by rpastrana** [comment]
@jpmcmu please review. 
The thought here is to cover known security exploits, even though codeQL ensures the proper APIs are used, this growing set of tests should confirm the proposed solutions actually behave as expected.

File: wsclient/src/test/java/org/hpccsystems/ws/client/security/XMLParsing.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        //AXP00010001: The parser has encountered more than "64000" entity expansions in this document; this is the limit imposed by the JDK.
+        Assert.fail("Unlimited XML entity expansion detected - should explicitely avoid and modern JDK should be targeted!");
+    }
+
+    @Test
+    public void noEntityExpansionTest() throws XPathExpressionException, ParserConfigurationException, SAXException, IOException 
+    {
+        try
+        {
+        File uploadFile = new File(millionLaughsFile);
```
Comment (jpmcmu) [inline wsclient/src/test/java/org/hpccsystems/ws/client/security/XMLParsing.java:111]: Formatting

**Review by jpmcmu — COMMENTED** [review]
@rpastrana a few questions

File: wsclient/src/test/java/org/hpccsystems/ws/client/security/XMLParsing.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            Assert.assertTrue("Unexpected SAXEception message detected, ensure DOCTYPE is disallowed and the feature \"http://apache.org/xml/features/disallow-doctype-decl\" set to true ", e.getMessage().equals("DOCTYPE is disallowed when the feature \"http://apache.org/xml/features/disallow-doctype-decl\" set to true."));
+        }
+    }
+
+
+    /*
+     * Illustrates security vulnerable XML parsing:
+     *
+     *@Test
+    public void unsafeExternalentityTest() throws XPathExpressionException, ParserConfigurationException, SAXException, IOException
```
[the remark this answers is not in the document]
Comment (rpastrana) [reply wsclient/src/test/java/org/hpccsystems/ws/client/security/XMLParsing.java:64]: we don't need to keep it, but chose to keep it to illustrate the wrong way- thought it'd be helpful

File: wsclient/src/test/java/org/hpccsystems/ws/client/security/XMLParsing.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    public void saferExternalentityTest() throws XPathExpressionException, ParserConfigurationException, SAXException, IOException 
+    {
+        try
+        {
+        File uploadFile = new File(externalEntityFile);
+        Utils.newSafeXMLDocBuilder().parse(uploadFile);
+        }
+        catch (SAXException e)
+        {
+            Assert.assertTrue("Unexpected SAXEception message detected, ensure DOCTYPE is disallowed and the feature \"http://apache.org/xml/features/disallow-doctype-decl\" set to true ", e.getMessage().equals("DOCTYPE is disallowed when the feature \"http://apache.org/xml/features/disallow-doctype-decl\" set to true."));
```
[the remark this answers is not in the document]
Comment (rpastrana) [reply wsclient/src/test/java/org/hpccsystems/ws/client/security/XMLParsing.java:93]: @jpmcmu please review latest changes.

**Comment by jpmcmu** [comment]
@rpastrana looks good

**Pull request merged**